File: benchmarks/r/b2_bubble_sort.R
# B2: bubble sort of 100000 LCG ints; output min and max. Scalar — slow tier.
# shared 64-bit LCG (seed 42) via 16-bit limbs — bit-identical to C/Python
.s0 <- 42; .s1 <- 0; .s2 <- 0; .s3 <- 0
.M0 <- 32557; .M1 <- 19605; .M2 <- 62509; .M3 <- 22609
.C0 <- 33103; .C1 <- 63335; .C2 <- 31614; .C3 <- 5125
.B16 <- 65536; .B32 <- 4294967296; .P15 <- 32768; .P33 <- 8589934592; .P31 <- 2147483648
next_double <- function() {
  p0 <- .s0*.M0; p1 <- .s0*.M1 + .s1*.M0
  p2 <- .s0*.M2 + .s1*.M1 + .s2*.M0
  p3 <- .s0*.M3 + .s1*.M2 + .s2*.M1 + .s3*.M0
  r0 <- p0 %% .B16; cy <- (p0 - r0)/.B16
  t1 <- p1 + cy; r1 <- t1 %% .B16; cy <- (t1 - r1)/.B16
  t2 <- p2 + cy; r2 <- t2 %% .B16; cy <- (t2 - r2)/.B16
  t3 <- p3 + cy; r3 <- t3 %% .B16
  u0 <- r0 + .C0; n0 <- u0 %% .B16; cy <- (u0 - n0)/.B16
  u1 <- r1 + .C1 + cy; n1 <- u1 %% .B16; cy <- (u1 - n1)/.B16
  u2 <- r2 + .C2 + cy; n2 <- u2 %% .B16; cy <- (u2 - n2)/.B16
  u3 <- r3 + .C3 + cy; n3 <- u3 %% .B16
  .s0 <<- n0; .s1 <<- n1; .s2 <<- n2; .s3 <<- n3
  L <- n2*.B32 + n1*.B16 + n0
  (n3*.P15 + (L - (L %% .P33))/.P33) / .P31
}
N <- 100000L
arr <- integer(N)
for (i in 1:N) arr[i] <- (next_double() * .P31) %% N
for (i in 1:(N-1)) {
  swapped <- FALSE
  for (j in 1:(N-i)) {
    if (arr[j] > arr[j+1]) { t <- arr[j]; arr[j] <- arr[j+1]; arr[j+1] <- t; swapped <- TRUE }
  }
  if (!swapped) break
}
cat(sprintf("%d %d\n", arr[1], arr[N]))

File: benchmarks/fortran/b2_bubble_sort.f90
! B2: bubble sort of 100000 ints — memory-access/cache behaviour.
! Values come from the shared 64-bit LCG (seed 42), matching the other
! LCG-based languages: arr(i) = (state>>33) mod N. Output: min and max.
program b2_bubble_sort
  implicit none
  integer, parameter :: N = 100000
  integer(8) :: state
  integer :: arr(N), i, j, tmp
  logical :: swapped
  state = 42_8
  do i = 1, N
     state = state * 6364136223846793005_8 + 1442695040888963407_8
     arr(i) = int(mod(ishft(state, -33), int(N, 8)), 4)
  end do
  do i = 1, N - 1
     swapped = .false.
     do j = 1, N - i
        if (arr(j) > arr(j+1)) then
           tmp = arr(j); arr(j) = arr(j+1); arr(j+1) = tmp; swapped = .true.
        end if
     end do
     if (.not. swapped) exit
  end do
  print '(I0,1X,I0)', arr(1), arr(N)
end program b2_bubble_sort
